File: src/main/java/org/raku/comma/annotation/fix/MakeSubmethodFix.java
package org.raku.comma.annotation.fix;

import com.intellij.codeInsight.intention.IntentionAction;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.util.IncorrectOperationException;
import org.raku.comma.psi.RakuElementFactory;
import org.raku.comma.psi.RakuRoutineDecl;
import org.jetbrains.annotations.NotNull;

public class MakeSubmethodFix implements IntentionAction {
    private final RakuRoutineDecl myDecl;

    public MakeSubmethodFix(RakuRoutineDecl decl) {
        myDecl = decl;
    }

    @NotNull
    @Override
    public String getText() {
        return getFamilyName();
    }

    @NotNull
    @Override
    public String getFamilyName() {
        return "Make submethod";
    }

    @Override
    public boolean isAvailable(@NotNull Project project, Editor editor, PsiFile file) {
        return true;
    }

    @Override
    public void invoke(@NotNull Project project, Editor editor, PsiFile file) throws IncorrectOperationException {
        PsiElement submethodDeclarator = RakuElementFactory.createRoutineDeclarator(project, "submethod");
        myDecl.getDeclaratorNode().replace(submethodDeclarator);
    }

    @Override
    public boolean startInWriteAction() {
        return true;
    }
}

File: src/main/java/org/raku/comma/inspection/fixes/MakeSubmethodFix.kt
package org.raku.comma.inspection.fixes

import com.intellij.codeInspection.LocalQuickFix
import com.intellij.codeInspection.ProblemDescriptor
import com.intellij.openapi.project.Project
import org.raku.comma.psi.RakuElementFactory

class MakeSubmethodFix(private val methodName: String) : LocalQuickFix {

    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
        val declarator = descriptor.psiElement
        val submethodDeclarator = RakuElementFactory.createRoutineDeclarator(project, "submethod")
        declarator.replace(submethodDeclarator)
    }

    override fun getName(): String { return "Convert '%s' to submethod".format(methodName) }
    override fun getFamilyName(): String { return "Make submethod" }
}